Implement `TypesIntersect(a, b *types.Type) bool`

ehalpern opened the pull request:
fixes: #3203

Commit:
Implement `TypesIntersect(a, b *types.Type) bool`

fixes: #3203

ehalpern commented:
fyi: @aboodman. @rafael-atticlabs, let me know if you'd like someone else to review.

rafael-atticlabs commented on go/types/intersect.go:
```diff
@@ -0,0 +1,92 @@
+	return typesIntersectImpl(a.Desc.(CompoundDesc).ElemTypes[0], b.Desc.(CompoundDesc).ElemTypes[0])
+}
+
+// TODO: consider requiring that both maps share a key type
```
agree. let's go ahead and do that.

rafael-atticlabs commented on go/types/intersect.go:
```diff
@@ -0,0 +1,92 @@
+	if aDesc.Name != bDesc.Name {
+		return false
+	}
+	for _, f := range aDesc.fields {
```
nit: let's avoid the map lookup here. desc.Fields is always sorted. do a parallel scan through both. i.e

```
for i, j := 0, 0; i < len(aDesc.fields) && j < len(bDesc.fields) {
  aName, bName := aDesc.fields[i].name, bDesc.fields[j].name;
  if aName < bName {
    i++
  } else if bName < aName {
    j++
  } else if typesIntersectImp(aDesc.fields[i].t, bDesc.fields[j].t) {
    return true
  }
  return false
}
```

rafael-atticlabs commented on go/types/intersect_test.go:
```diff
@@ -0,0 +1,88 @@
+	"github.com/attic-labs/testify/assert"
+)
+
+func TestIntersectImpl(t *testing.T) {
```
add some tests which cover hit & miss cases for containers with union parameters. e.g. List<A|B|C> & List<C|D>.

Also maybe one "recursive" struct case, like {a: List<B>} & {a: List<A|B>}

Commit:
Respond to review:
- Require at least one key type to match for map intersction
- Do parallel scan of field when comparing struct
- Add more tests for unions and nested collections

ehalpern commented:
PTAL @rafael-atticlabs

arv commented:
How is this different from IsSubtype?

rafael-atticlabs commented:
@arv 
```
TypesIntersect(Struct { a: Number, b: String }, Struct { a: Number, c: Bool }) == true
```

arv commented:
Can we use another word to describe this. Type intersection is already a well defined term in type theory.

aboodman commented:
Note that this doesn't return a new type... it just returns whether the two
types do intersect or not. Is this not type intersection from theory?
On Fri, Feb 24, 2017 at 7:37 AM Erik Arvidsson <notifications@github.com>
wrote:

> Can we use another word to describe this. Type intersection is already a
> well defined term in type theory.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/attic-labs/noms/pull/3223#issuecomment-282321733>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAE6BHIfiCtCgKC-QNimUah1SFfLF9EDks5rfvk0gaJpZM4MKmgO>
> .
>

arv commented:
You can always get the intersection of two different types but there might not be any values that fit in the type intersection.

```
type T = number & string;
const n: T = ???;
```

There is no value that can be a number and a string....

```
type T2 = {b: string} & {a: number};
const n2: T2 = {a: 42, b: 'a'}
```

Given this it is not clear what it means that `{a: number, b: string} & {b: string, c: boolean}` returns `true`? Maybe the question we are asking is if there is a common field in two types?

arv commented:
I think the terminology needs some tweaking. The question we are asking is the type may help get us towards the other type. In other words if there is a common (non trivial) super type 

containsCommonSuperType
mayContribute{To,Towards}

rafael-atticlabs commented:
Per meatspace discussion with @arv, let's change the name of this to be `ContainCommonSupertype(a, b *types.Type) bool`

aboodman commented:
I like that name a lot!

On Fri, Feb 24, 2017 at 10:18 AM, Rafael Weinstein <notifications@github.com
> wrote:

> Per meatspace discussion with @arv <https://github.com/arv>, let's change
> the name of this to be ContainCommonSupertype(a, b *types.Type) bool
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/attic-labs/noms/pull/3223#issuecomment-282364199>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAE6BK8id_UT8f-w3VL26bK_RUhvY_Hhks5rfx8EgaJpZM4MKmgO>
> .
>

aboodman commented:
Except for it it's not grammatical. Should be ContainsCommonSupertype
(plural)

On Fri, Feb 24, 2017 at 10:30 AM, Aaron Boodman <aa@attic.io> wrote:

> I like that name a lot!
>
> On Fri, Feb 24, 2017 at 10:18 AM, Rafael Weinstein <
> notifications@github.com> wrote:
>
>> Per meatspace discussion with @arv <https://github.com/arv>, let's
>> change the name of this to be ContainCommonSupertype(a, b *types.Type)
>> bool
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/attic-labs/noms/pull/3223#issuecomment-282364199>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAE6BK8id_UT8f-w3VL26bK_RUhvY_Hhks5rfx8EgaJpZM4MKmgO>
>> .
>>
>
>

ehalpern commented:
@aboodman: Sounds good, but I think ContainCommonSupertype is correct since it refers to two arguments (i.e. a and b contain common supertypes)

rafael-atticlabs commented:
there are two things you are asking the question about. e.g. "a and b contain a common supertype".

Commit:
Rename to ContainCommonSupertype

ehalpern commented:
Done

aboodman commented:
You are right! My mistake.
